pkg/util/hlc/hlc.go

```diff
@@ -125,6 +155,164 @@ func NewClock(physicalClock func() int64, maxOffset time.Duration) *Clock {
+// jumps between two successive reads of the physical clock.
+//
+// This should only be called once per clock, and will return an error if called
+// more than once
+//
+// tickerFn is used to create a new ticker
+//
+// tickCallback is called whenever maxClockJumpCh or a ticker tick is processed
+func (c *Clock) StartMonitoringClockJumps(
+	maxClockJumpCh <-chan time.Duration,
+	tickerFn func(d time.Duration) *time.Ticker,
+	tickCallback func(),
+) error {
+	alreadyMonitoring := c.setMonitoringClockJump()
+	if alreadyMonitoring {
+		return errors.New("clock jumps are already being monitored")
+	}
+
+	go func() {
+		ticker := tickerFn(c.maxOffset)
+		ticker.Stop()
+		for {
+			select {
+			case maxClockJump, ok := <-maxClockJumpCh:
+				ticker.Stop()
+				if !ok {
+					return
+				}
+
+				refreshLastUpdateTimeItvl := maxClockJump / 2
+				if refreshLastUpdateTimeItvl > 0 {
+					ticker = tickerFn(refreshLastUpdateTimeItvl)
+				}
+				c.updateMaxClockJump(maxClockJump)
+			case <-ticker.C:
+				c.PhysicalNow()
+			}
+
+			if tickCallback != nil {
+				tickCallback()
+			}
+		}
+	}()
+	return nil
+}
+
+// wallTimeToPersist returns the wall time offset by persistFutureWallTimeDelta
+func (c *Clock) wallTimeToPersist() int64 {
+	return c.Now().WallTime + persistFutureWallTimeDelta
+}
+
+// EnsureInitialWallTimeMonotonicity sleeps till the wall time reaches
+// prevWallTime. prevWallTime > 0 implies we need to guarantee wall time
+// monotonicity across server restarts. prevWallTime is the last successfully
+// persisted timestamp greater then any wall time used by the server.
+//
+// persistWallTimeFn is used to persist the future wall time, and should return
+// an error if the persist fails.
+func (c *Clock) EnsureInitialWallTimeMonotonicity(
+	ctx context.Context,
+	prevWallTime int64,
+	persistWallTimeFn func(int64) error,
+	sleepFn func(d time.Duration),
+) error {
+	for {
+		delta := time.Duration(prevWallTime-c.Now().WallTime) + 1
+		if delta <= 0 {
+			break
+		}
+
+		log.Infof(ctx, "Sleeping for %v to catch up to wall clock time of %v",
+			delta, prevWallTime)
+		sleepFn(delta)
+	}
+
+	if prevWallTime > 0 {
+		// A non zero prevWallTime signifies that the feature to persist future
+		// wall times is enabled. Persist a new future wall time to continue
+		// guaranteeing monotonicity
+		return c.updateFutureWallTime(persistWallTimeFn, c.wallTimeToPersist())
+	}
+	return nil
+}
+
+// StartPersistingFutureWallTime starts a goroutine to persist a future wall time.
+// The values pushed in persistFutureWallTimeCh specify whether the future
+// timestamp should be persisted.
+//
+// persistWallTimeFn is used to persist the future wall time, and should return
+// an error if the persist fails
+//
+// tickerFn is used to create a new ticker
+//
+// tickCallback is called whenever maxClockJumpCh or a ticker tick is processed
+func (c *Clock) StartPersistingFutureWallTime(
```

Comment:
I find the approach of using a channel to read config value, and resetting a timer on every tick heavy handed for this usage. I would have just had a loop that would call time.Sleep where the sleep time is determined by the current value of futureTimeDelta. If futureTimeDelta is disabled, it would not persist and just sleep for 1 second. And I would have optionally considered launching the goroutine only when futureTimeDelta is enabled which can be done in the callback. I would not bother with disabling the goroutine if the setting is disabled online.

Reply:
There are some advantages of resetting the timer.
1. If `futureTimeDelta` is disabled, then there is no polling. The goroutine blocks on the channel till this value is changed. So there are no context switches.
2. If we set `futureTimeDelta` to a large value and then to a small value, the timer based approach will immediately switch to the small value. The sleep based approach will only switch after it wakes up (which could be a long time if `futureTimeDelta` is accidentally set to a large value).
3. Only a single goroutine is dedicated for this. There is no chance of goroutine leak.

If we launch a new goroutine on every setting change, there could be a risk of a goroutine leak.

Reply:
My suggestion is that if futureTimeDelta is disabled, then no goroutine be launched until it is enabled. Once launched, this goroutine is not stopped and not replaced with another goroutine until the process exits. There is no overhead of an extra goroutine if the setting is always disabled, and no risk of any goroutine leaks. The extra periodic check at every 1 second will occur only when the setting is enabled and then disabled online. The performance impact of such a goroutine will be negligible by any standards. I bet that the cost of creating a new timer on every tick in the common case would be much more than that. And we can always set a maximum value of 1 second on the sleep regardless of what the config says (of course, the value of the persisted timestamp will be based on the config).

Reply:
> My suggestion is that if futureTimeDelta is disabled, then no goroutine be launched until it is enabled. 

In cluster start-up, we don't have access to see if the cluster setting for `futureTimeDelta` is enabled because cluster settings are not initialized. We decide this on the basis of whether any value has been persisted for `futureTimeDelta`. I see no harm in starting a goroutine and blocking it on a channel. Starting a goroutine the first time this cluster setting is enabled / disabled would need to have synchronisation adding some extra complexity

> The extra periodic check at every 1 second will occur only when the setting is enabled and then disabled online

Yes, but this can be still be avoided.

>I bet that the cost of creating a new timer on every tick in the common case would be much more than that.

A new timer is only created if the cluster setting changes (which will be rare), not on every tick

Reply:
> In cluster start-up, we don't have access to see if the cluster setting for futureTimeDelta is enabled because cluster settings are not initialized.

I don't see how this will add more complexity but I could be missing something. What I'm suggesting would essentially amount to keeping an indicator variable in Clock.mu to track whether the time persistence goroutine is currently active, and the callback in SetOnChange looks at this indicator variable and launches a goroutine if one does not already exist. As far as I can tell, the behavior would be equivalent to what the current code does.

> Yes, but this can be still be avoided.

I would not justify extra code complexity based on avoiding negligible performance impact which cannot be measured.

> A new timer is only created if the cluster setting changes (which will be rare), not on every tick

Right, I got this wrong when I read it first. But this was not the main point in my argument.

Reply:
The earlier point is still valid
> If we set futureTimeDelta to a large value and then to a small value, the channel + timer based approach will immediately switch to the small value. The sleep based approach will only switch after it wakes up (which could be a long time if futureTimeDelta is accidentally set to a large value).

If we are using the channel based approach, it's simple to just have a sleeping goroutine whenever this cluster setting is disabled.

Reply:
> If we set futureTimeDelta to a large value and then to a small value, the channel + timer based approach will immediately switch to the small value. The sleep based approach will only switch after it wakes up (which could be a long time if futureTimeDelta is accidentally set to a large value).

I think this can be overcome by enforcing a maximum sleep time of 1 second or something.

>  If we are using the channel based approach, it's simple to just have a sleeping goroutine whenever this cluster setting is disabled.

I think it is better to not launch the goroutine at all when the setting is disabled from the point of view of safety and risk of bugs. This doesn't apply to Rubrik because we would run with the setting enabled, but might be relevant to the common case of cockroach deployment.

Reply:
> I think this can be overcome by enforcing a maximum sleep time of 1 second or something.

I still prefer to block on a channel compared to polling. No unnecessary context switches.

> I think it is better to not launch the goroutine at all when the setting is disabled from the point of view of safety and risk of bugs. This doesn't apply to Rubrik because we would run with the setting enabled, but might be relevant to the common case of cockroach deployment.

The launched goroutine will be immediately blocked on reading from a channel. We can check with CRDB folks if they see a problem in this. If they also prefer to not launch the goroutine we launch it first time it is set to true. I would prefer to have to same behaviour on whether the server started with this setting as `false` and if this is set to `false` from `true` while the server is running.